### PR TITLE
R6 Methods in FunctionReporter

### LINKED ISSUE #123: FunctionReporter handling of R6 class methods
`FunctionReporter` does not currently inspect R6 classes for their methods. For packages that lean heavily on R6 classes, this can be a significant part of the function dependency network that is effectively missing.

Note that this is different issue than #14 which is around graphing the inheritance network between classes.

### PR DESCRIPTION
**Pull request by jayqi**
Resolves #123 

Summary of changes:
- Added support for R6 class methods in `FunctionReporter`. Some design choices/constraints:
    - Class names are the _name of the generator object in the package namespace_, and _not_ the `classname` attribute. This is because the `classname` attribute is not required to be defined and also not required to match the generator object name, though people conventionally do assign one with the same name. 
    - Nodes have names of the form `<classname>$<methodtype>$<methodname>`, e.g., `FunctionReporter$private_methods$extract_nodes `. 
    - Methods only ever show up as one node even with inheritance, unless the child class overrides with a new definition. Inherited methods are assigned to the class in which they were defined, and not for any child classes
    - Dependency on calls to instantiated objects' methods are not supported. Instantiated objects can have arbitrary names and I don't want to deal with it right now. 
- Added test package `milne` for testing R6 classes
- ~Changed the .gitignore entries for data files to only exclude data files in package root. Want to be able to commit test data in tests.~ Added .gitignore exclusion for csv files in `tests/testthat/testdata` directory. 

Not addressed in this PR. Maybe for discussion in issues after merging. 
- Calls to object constructors by functions are ignored right now, like `FunctionReporter$new()` in `DefaultReporters`? Not clear the right way to represent these. Does that edge link to the `initialize` method for that class? `FunctionReporter` has no `initialize` method, so would we link the edge to the `initialize` method inherited up the ancestor sequence? But actually none of its ancestors have an `initialize` method defined -- they're all using the vanilla one built-in to R6. So there isn't even a node right now for it.

### CODE REVIEW

File: inst/milne/R/The_End.R [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+)
+
+#' @title Age Four
+#' @name Four
+#' @family TheEnd
+#' @description Age Four
+#' @importFrom R6 R6Class
+#' @export
+Four <- R6::R6Class(
+    # R6 classes don't need classname at all
```
Comment (bburns632) [inline inst/milne/R/The_End.R]: I did not know this was possible.  TIL.

File: R/PackageFunctionReporter.R [hunk 2]
```diff
@@ -175,7 +228,18 @@ FunctionReporter <- R6::R6Class(
             })
             exported_obj_names <- ls(sprintf("package:%s", self$pkg_name))
             nodes[, isExported := node %in% exported_obj_names]
-
+            
+            # Check if we have R6 functions
+            if (length(self$pkg_R6_classes) > 0) {
+                r6DT <- self$pkg_R6_methods[, .(
+                    node = paste(CLASS_NAME, METHOD_TYPE, METHOD_NAME, sep = "$")
```
[the remark this answers is not in the document]
Comment (jayqi) [reply R/PackageFunctionReporter.R]: `self$pkg_R6_methods` is a data.table.

File: R/PackageFunctionReporter.R [hunk 3]
```diff
@@ -265,17 +352,266 @@ FunctionReporter <- R6::R6Class(
...
+#' @importFrom data.table rbindlist
+.get_R6_class_inheritance <- function(class_names, pkg_name, pkg_env) {
+    inheritanceDT <- data.table::rbindlist(lapply(
+        X = class_names
+        , FUN = function(x, p = pkg_env) {
+            parentClassName <- deparse(get(x, p)$inherit)
+            parentClassGenerator <- get(x, p)$get_inherit()
+            return(list(
+                CLASS_NAME = x
+                , PARENT_NAME = if (!is.null(parentClassGenerator)) {
```
[the remark this answers is not in the document]
Comment (jayqi) [reply R/PackageFunctionReporter.R]: It's also inconsistent between the nodes and the edges. I kept the within-table convention? Haha. Future PR? `:shrug:`

**Comment by jayqi** [comment]
@jameslamb Addressed your comments! Thanks for reviewing.

File: R/PackageFunctionReporter.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply R/PackageFunctionReporter.R]: ahhhh I see now

File: inst/milne/R/The_End.R (hunk 1 above)
Comment (jameslamb) [reply inst/milne/R/The_End.R]: same, thanks for teaching us @jayqi

File: R/PackageFunctionReporter.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply R/PackageFunctionReporter.R]: fair fair fair fair fair

**Comment by jameslamb** [comment]
This is good with me! Will leave it to @bburns632 to do one more pass and and smash merge, since this is a decent-sized PR

**Comment by bburns632** [comment]
@jayqi this PR brings to light the bottom depths of an R package.  It is the next level of insight into a package.  Excellent work. 

AND I look forward to PR #129 to add some structure to these newly surfaced elements.  

Here's master currently for a `pkgnet` report on `pkgnet`: 
 
image

Here's the same report with changes from this PR: 
image

**Pull request merged**